False positive CA1508: '{0}' is always 'null'. Remove or refactor the condition(s) to avoid dead code.

**Version Used**: 
Microsoft Visual Studio Enterprise 2019 Preview
Version 16.10.0 Preview 1.0
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>

</Project>
```

**Steps to Reproduce**:
```csharp
namespace Ca1508FalsePositive
{
    using System;
    using System.Collections.Generic;

    class Program
    {
        private static object AutoDetectAndConnect(string host)
        {
            List<Exception> exceptions = null;
            foreach (string registeredScheduler in s_schedulers)
            {
                try
                {
                    object conn = Connect(registeredScheduler, host);
                    if (conn != null)
                        return conn;
                }
                catch (Exception ex)
                {
                    // warning CA1508: 'exceptions' is always 'null'. Remove or refactor the condition(s) to avoid dead code.
                    (exceptions ??= new List<Exception>()).Add(ex);
                }
            }

            // Commenting out this code block will "resolve" CA1508 above
            if (exceptions == null)
            {
                throw new NotSupportedException(host);
            }

            throw new AggregateException(exceptions);

            static object Connect(string a, string b) => throw new NotImplementedException($"{a}{b}");
        }

        static List<string> s_schedulers = new List<string>();
        static void Register(string scheduler) => s_schedulers.Add(scheduler);

        static void Main(string[] args)
        {
            Register("abc");
            Register("def");

            try
            {
                var conn = AutoDetectAndConnect(args.Length > 0 ? args[0] : args.GetHashCode().ToString());
                Console.WriteLine(conn);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
        }
    }
}
```

**Expected Behavior**:
I should not be warning that `exceptions` is always `null`. It is only `null` with the first exception encountered.

**Actual Behavior**:
I receive CA1508 when lazily allocating the list. Interestingly enough, I do not receive CA1508 when I comment out the check for a null list below.

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -7762,5 +7762,73 @@ internal class SerialPort
                 LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
             }.RunAsync();
         }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.PointsToAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Fact, WorkItem(4984, "https://github.com/dotnet/roslyn-analyzers/issues/4984")]
+        public async Task NullCoalescingOperatorInsideCatchBlock_WithinLoop_NoDiagnosticsAsync()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = @"
+namespace Ca1508FalsePositive
+{
+    using System;
+    using System.Collections.Generic;
+
+    class Program
+    {
+        private static object AutoDetectAndConnect(string host)
+        {
+            List<Exception> exceptions = null;
+            foreach (string registeredScheduler in s_schedulers)
+            {
+                try
+                {
+                    object conn = Connect(registeredScheduler, host);
+                    if (conn != null)
+                        return conn;
+                }
+                catch (Exception ex)
+                {
+                    // warning CA1508: 'exceptions' is always 'null'. Remove or refactor the condition(s) to avoid dead code.
+                    (exceptions ??= new List<Exception>()).Add(ex);
+                }
+            }
+
+            // Commenting out this code block will ""resolve"" CA1508 above
+            if (exceptions == null)
+            {
+                throw new NotSupportedException(host);
+            }
+
+            throw new AggregateException(exceptions);
+
+            static object Connect(string a, string b) => throw new NotImplementedException($""{a}{b}"");
+        }
+
+        static List<string> s_schedulers = new List<string>();
+        static void Register(string scheduler) => s_schedulers.Add(scheduler);
+
+        static void Main(string[] args)
+        {
+            Register(""abc"");
+            Register(""def"");
+
+            try
+            {
+                var conn = AutoDetectAndConnect(args.Length > 0 ? args[0] : args.GetHashCode().ToString());
+                Console.WriteLine(conn);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
+            }
+        }
+    }
+}",
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
+            }.RunAsync();
+        }
     }
 }

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs
@@ -7696,5 +7696,71 @@ internal class C
 
             await test.RunAsync();
         }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.PointsToAnalysis)]
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.NullAnalysis)]
+        [Fact, WorkItem(5199, "https://github.com/dotnet/roslyn-analyzers/issues/5199")]
+        public async Task InvocationToExceptionThrowingMethod_InsideTryBlock_NoDiagnosticsAsync()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = @"
+#nullable enable
+
+using System;
+
+class C
+{
+    public SerialPort? OpenPort2(string portName)
+    {
+        SerialPort? tempPort = null;
+        SerialPort? port = null;
+        try
+        {
+            tempPort = new SerialPort(portName);
+            tempPort.Open();
+            SomeMethod();
+            port = tempPort;
+            tempPort = null;
+
+        }
+        finally
+        {
+            if (tempPort != null) // BOOM
+            {
+                tempPort.Close();
+            }
+        }
+        return port;
+    }
+
+    private void SomeMethod()
+    {
+        throw new NotImplementedException();
+    }
+}
+
+internal class SerialPort
+{
+    private string portName;
+
+    public SerialPort(string portName)
+    {
+        this.portName = portName;
+    }
+
+    internal void Close()
+    {
+        throw new NotImplementedException();
+    }
+
+    internal void Open()
+    {
+        throw new NotImplementedException();
+    }
+}",
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
+            }.RunAsync();
+        }
     }
 }

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs
@@ -3306,5 +3306,77 @@ class C
 }",
             }.RunAsync();
         }
+
+        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.ValueContentAnalysis)]
+        [Fact, WorkItem(5160, "https://github.com/dotnet/roslyn-analyzers/issues/5160")]
+        public async Task CatchBlock_WithinForLoop_NoDiagnosticsAsync()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = @"
+using System;
+
+class C
+{
+    void M1()
+    {
+        const int attempts = 5;
+
+        for (int i = 0; i < attempts; i++)
+        {
+            try
+            {
+                Console.WriteLine($""Hello world: {i}"");
+
+                if (i == attempts - 1) // Last iteration.
+                {
+                    throw new InvalidOperationException(""Oops, something went wrong!"");
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($""Exception caught: {ex.Message}"");
+
+                if (i == attempts - 1)
+                {
+                    Console.WriteLine(""This should never happen, according to CA1508."");
+                }
+            }
+        }
+    }
+
+    void M2()
+    {
+        for (int i = 0; i <= 5; i++)
+        {
+            try
+            {
+                MayThrowException(i);
+                if (i == 0)
+                {
+                    throw new Exception();
+                }
+            }
+            catch (Exception)
+            {
+                if (i == 1)
+                {
+                    Console.WriteLine();
+                }
+            }
+        }
+    }
+
+    private void MayThrowException(int i)
+    {
+        if (i % 2 == 1)
+        {
+            throw new Exception();
+        }
+    }
+}",
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp8,
+            }.RunAsync();
+        }
     }
 }

--- a/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerationsTests.cs
@@ -181,7 +181,7 @@ End Namespace
             await VerifyVisualBasicAsync(vbCode);
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/roslyn-analyzers/issues/6830")]
         public async Task TestInvocationsInForEachLoop()
         {
             var csharpCode = @"
@@ -4067,7 +4067,7 @@ End Module
                 customizedLinqChainMethods: editorConfig);
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/roslyn-analyzers/issues/6830")]
         public async Task TestDomainMerge()
         {
             var csharpCode1 = @"

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/CopyAnalysis/CopyAnalysis.CopyDataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/CopyAnalysis/CopyAnalysis.CopyDataFlowOperationVisitor.cs
@@ -397,9 +397,9 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.CopyAnalysis
                         continue;
                     }
 
+                    var existingValue = targetAnalysisData[key];
                     if (CurrentAnalysisData.TryGetValue(key, out var newValue))
                     {
-                        var existingValue = targetAnalysisData[key];
                         if (newValue.AnalysisEntities.Count == 1)
                         {
                             if (existingValue.AnalysisEntities.Count == 1)
@@ -422,6 +422,23 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.CopyAnalysis
                         }
 
                         processedEntities.AddRange(newValue.AnalysisEntities);
+                    }
+                    else
+                    {
+                        var entitiesToExclude = existingValue.AnalysisEntities.Where(e => !CurrentAnalysisData.HasAbstractValue(e));
+                        var excludedCount = 0;
+                        foreach (var entity in entitiesToExclude)
+                        {
+                            targetAnalysisData.RemoveEntries(entity);
+                            processedEntities.Add(entity);
+                            excludedCount++;
+                        }
+
+                        if (excludedCount < existingValue.AnalysisEntities.Count)
+                        {
+                            newValue = existingValue.WithEntitiesRemoved(entitiesToExclude);
+                            targetAnalysisData.SetAbstactValueForEntities(newValue, entityBeingAssigned: null);
+                        }
                     }
                 }
             }

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowAnalysis.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowAnalysis.cs
@@ -86,7 +86,7 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
             // i.e. for every '{key, value}' pair in the dictionary, 'key' is the destination of at least one back edge
             // and 'value' is the minimum ordinal such that there is no back edge to 'key' from any basic block with ordinal > 'value'.
             using var loopRangeMap = PooledDictionary<int, int>.GetInstance();
-            ComputeLoopRangeMap(cfg, loopRangeMap);
+            var hasAnyTryBlock = ComputeLoopRangeMap(cfg, loopRangeMap);
 
             TAnalysisData? normalPathsExitBlockData = null, exceptionPathsExitBlockData = null;
 
@@ -115,7 +115,12 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
                     blockToUniqueInputFlowMap, loopRangeMap, exceptionPathsAnalysisPostPass: false);
                 normalPathsExitBlockData = resultBuilder.ExitBlockOutputData;
 
-                if (analysisContext.ExceptionPathsAnalysis)
+                // If we are executing exception paths analysis OR have at least one try/catch/finally block
+                // in the method, execute an exception path analysis post pass.
+                // This post pass will handle all possible operations within the control flow graph that can
+                // throw an exception and merge analysis data after all such operation analyses into the
+                // catch blocks reachable from those operations.
+                if (analysisContext.ExceptionPathsAnalysis || hasAnyTryBlock)
                 {
                     RoslynDebug.Assert(normalPathsExitBlockData != null);
 
@@ -286,11 +291,9 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
 
                 // Check if we are starting a try region which has one or more associated catch/filter regions.
                 // If so, we conservatively merge the input data for try region into the input data for the associated catch/filter regions.
-#pragma warning disable CA1508 // Avoid dead conditional code - https://github.com/dotnet/roslyn-analyzers/issues/4408
                 if (block.EnclosingRegion?.Kind == ControlFlowRegionKind.Try &&
                     block.EnclosingRegion.EnclosingRegion?.Kind == ControlFlowRegionKind.TryAndCatch &&
                     block.EnclosingRegion.EnclosingRegion.FirstBlockOrdinal == block.Ordinal)
-#pragma warning restore CA1508 // Avoid dead conditional code
                 {
                     MergeIntoCatchInputData(block.EnclosingRegion.EnclosingRegion, input, block);
                 }
@@ -791,14 +794,19 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
             }
         }
 
-        private static void ComputeLoopRangeMap(ControlFlowGraph cfg, PooledDictionary<int, int> loopRangeMap)
+        private static bool ComputeLoopRangeMap(ControlFlowGraph cfg, PooledDictionary<int, int> loopRangeMap)
         {
+            var hasAnyTryBlock = false;
             for (int i = cfg.Blocks.Length - 1; i > 0; i--)
             {
                 var block = cfg.Blocks[i];
                 HandleBranch(block.FallThroughSuccessor);
                 HandleBranch(block.ConditionalSuccessor);
+
+                hasAnyTryBlock |= block.EnclosingRegion.Kind == ControlFlowRegionKind.Try;
             }
+
+            return hasAnyTryBlock;
 
             void HandleBranch(ControlFlowBranch? branch)
             {

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
@@ -923,7 +923,6 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
 
         protected virtual TAnalysisData GetMergedAnalysisDataForPossibleThrowingOperation(TAnalysisData? existingData, IOperation operation)
         {
-            Debug.Assert(DataFlowAnalysisContext.ExceptionPathsAnalysis);
             Debug.Assert(ExecutingExceptionPathsAnalysisPostPass);
 
             return existingData == null ?

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
@@ -844,7 +844,6 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
 
         protected virtual void HandlePossibleThrowingOperation(IOperation operation)
         {
-            Debug.Assert(DataFlowAnalysisContext.ExceptionPathsAnalysis);
             Debug.Assert(ExecutingExceptionPathsAnalysisPostPass);
 
             // Bail out if we are not analyzing an interprocedural call and there is no

